### ISSUE TITLE
FullAjaxExceptionHandler does not show error page when exception is thrown during PrimeFaces 11 ajax request

### ISSUE THREAD

**Issue by crick77** [omnifaces/omnifaces#686]
Hello everyone during developmento of a JSF 2 application i stumbled on some weird behavior of OmniFaces FullAjaxExecptionHandler. When an ajax request throws an error (RuntimeException descendands) the error page is not displayed. I tried to remove everything from web.xml and faces-config.xml making them as much as similar to the OmniFaces's demo showcase ones (https://github.com/omnifaces/showcase/tree/master/src/main/webapp/WEB-INF). Here is my web.xml:

`<?xml version="1.0" encoding="UTF-8"?>
<web-app version="4.0" xmlns="http://xmlns.jcp.org/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_4_0.xsd">
    <context-param>
        <param-name>javax.faces.PROJECT_STAGE</param-name>
        <param-value>Development</param-value>
    </context-param>    
    <context-param>
        <param-name>javax.faces.FACELETS_BUFFER_SIZE</param-name>
        <param-value>65535</param-value> <!-- 64KB. -->
    </context-param>
    <context-param>
        <param-name>javax.faces.FACELETS_SKIP_COMMENTS</param-name>
        <param-value>true</param-value>
    </context-param>
    <context-param>
        <param-name>javax.faces.INTERPRET_EMPTY_STRING_SUBMITTED_VALUES_AS_NULL</param-name>
        <param-value>true</param-value> <!-- https://balusc.omnifaces.org/2015/10/the-empty-string-madness.html -->
    </context-param>
    <context-param>
        <param-name>javax.faces.SERIALIZE_SERVER_STATE</param-name>
        <param-value>true</param-value> <!-- Mojarra defaults to false and MyFaces to true. Let's align out.  -->
    </context-param>
    <servlet>
        <servlet-name>facesServlet</servlet-name>
        <servlet-class>javax.faces.webapp.FacesServlet</servlet-class>
        <load-on-startup>1</load-on-startup>
    </servlet>
    <servlet-mapping>
        <servlet-name>facesServlet</servlet-name>
        <url-pattern>*.xhtml</url-pattern>
    </servlet-mapping>
    <filter>
        <filter-name>facesExceptionFilter</filter-name>
        <filter-class>org.omnifaces.filter.FacesExceptionFilter</filter-class>
    </filter>
    <filter-mapping>
        <filter-name>facesExceptionFilter</filter-name>
        <servlet-name>facesServlet</servlet-name>
    </filter-mapping>                                                                                                                                                                                                           
    <session-config>
        <session-timeout>
            30
        </session-timeout>
    </session-config>
    <error-page>
        <exception-type>java.lang.RuntimeException</exception-type>
        <location>/WEB-INF/errorpages/500.xhtml</location>
    </error-page>  
    <error-page>
        <error-code>500</error-code>
        <location>/WEB-INF/errorpages/500.xhtml</location>
    </error-page>    
    <error-page>
        <error-code>404</error-code>
        <location>/WEB-INF/errorpages/404.xhtml</location>
    </error-page>
    <welcome-file-list>
        <welcome-file>index.xhtml</welcome-file>
    </welcome-file-list>
    <mime-mapping>
        <extension>eot</extension>
        <mime-type>application/vnd.ms-fontobject</mime-type>
    </mime-mapping>
    <mime-mapping>  
        <extension>otf</extension>  
        <mime-type>font/opentype</mime-type>  
    </mime-mapping>      
    <mime-mapping>  
        <extension>ttf</extension>  
        <mime-type>application/x-font-ttf</mime-type>  
    </mime-mapping>      
    <mime-mapping>  
        <extension>woff</extension>  
        <mime-type>application/x-font-woff</mime-type>  
    </mime-mapping>
    <mime-mapping>  
        <extension>svg</extension>  
        <mime-type>image/svg+xml</mime-type>  
    </mime-mapping>
    <mime-mapping>  
        <extension>woff2</extension>  
        <mime-type>application/x-font-woff2</mime-type>  
    </mime-mapping>
</web-app>`

Then the faces-config.xml:

`<faces-config version="2.3"
              xmlns="http://xmlns.jcp.org/xml/ns/javaee"
              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-facesconfig_2_3.xsd">            
    <factory>
        <exception-handler-factory>org.omnifaces.exceptionhandler.FullAjaxExceptionHandlerFactory</exception-handler-factory>
    </factory>                
</faces-config>`

in a bean, just to do some test i've put a simple action:

`@Named
@RequestScoped
public class ErrorController extends BaseController {
    public void error() {
        throw new NumberFormatException("This is an error!");
    }
}`

and in the page.xhtml

`<p:commandButton id="error" value="Send error!" action="#{errorController.error()}" />`

In the appserver log the FullAjaxExceptionHandler prints this:

`[5985dfe0-452a-4d1d-a17f-89008ee1801d][0:0:0:0:0:0:0:1] FullAjaxExceptionHandler: An exception occurred during processing JSF ajax request. Error page '/WEB-INF/errorpages/500.xhtml' will be shown.
java.lang.NumberFormatException: This is an error!
    at it.usr.web.estrazioni.controller.ErrorController.error(ErrorController.java:19)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    ...`

but the page is not displayed, instead the same view is redisplayed but with some css/styles not applied due to some missing resources. This problem does not occur when i put `ajax="false"` on `p:commandButton`, but this is an expected behavior. When and ajax request is sent the redirection to errorpage does not work.
My current runtime environment is:

- Java 8.0u291 64bit
- Payara 5.2021.2 (tried also with 5.2021.7)
- Java EE 8 environment
- PrimeFaces 11.0.0
- OmniFaces 3.13

PS. I have a couple of filters that I've removed for testing purposes. First filter is an authorization filter (a slightly modified version of BalusC one) and a LogFilter that extends org.omnifaces.filter.FacesExceptionFilter that only outputs some more detailed log in a logfile. They were both applied to the facesServlet. The strange thing, in this case, is that the method logException() is not even called! The filter is initialized and Logger injection works (the property is non-null).

Thank you for your help.

**Comment by melloware** [comment]
Can you put together a small reproducer using  [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project?

I think it will help the developers see and solve the issue.

**Comment by BalusC** [comment]
Reproduced. Fails in PF 11.0.0 but works (half) in PF 10.0.0. 

Partial response of FullAjaxExceptionHandler (which acts like `update="@all"`) contains `<update>` elements for `javax.faces.ViewRoot` and `javax.faces.ViewState`.

---

**PF 11.0.0**

Update of `javax.faces.ViewRoot`: fail during `$("body").html(...)` call in `updateBody` function
Update of `javax.faces.ViewState`: didn't got that far

Stack trace of failing `$("body").html(...)` call:
```
TypeError: Cannot read properties of null (reading 'detachedWidgets')
    at HTMLButtonElement.<anonymous> (http://localhost:8080/playground-tomcat9/javax.faces.resource/core.js.xhtml?ln=primefaces&v=11.0.0:24:1033)
    at HTMLButtonElement.dispatch (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:43064)
    at HTMLButtonElement.v.handle (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:41048)
    at Object.trigger (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:71515)
    at S.fn.init.triggerHandler (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:72194)
    at Function.cleanData (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=11.0.0:14:219)
    at S.fn.init.empty (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:50317)
    at S.fn.init.<anonymous> (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:50808)
    at $ (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:32425)
    at S.fn.init.html (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=11.0.0:2:50494)
```
Thrown by following block in `BaseWidget#init()` because `window.PrimeFaces` was cleared out to `null` during update of `javax.faces.ViewRoot`.
```
                this.jq.on("remove", function() {
                    PrimeFaces.detachedWidgets.push(b.widgetVar)
                })
```

Result: body is not updated at all (matches your issue report).

---

**PF 10.0.0**

Update of `javax.faces.ViewRoot`: success
Update of  `javax.faces.ViewState`: fail during accessing `PrimeFaces.ajax` in below block because `window.PrimeFaces` was cleared out to `null` during update of `javax.faces.ViewRoot`.
```
case "update":
    PrimeFaces.ajax.ResponseProcessor.doUpdate(l, m, b);
    break;
```
Stack trace of failing `PrimeFaces.ajax` access:
```
TypeError: Cannot read properties of null (reading 'ajax')
    at Object.handle (http://localhost:8080/playground-tomcat9/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0:20:13088)
    at Object.<anonymous> (http://localhost:8080/playground-tomcat9/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0:20:9896)
    at c (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0:2:28294)
    at Object.fireWith [as resolveWith] (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0:2:29039)
    at l (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0:2:79800)
    at XMLHttpRequest.<anonymous> (http://localhost:8080/playground-tomcat9/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0:2:82254)
```
Result: body is updated but view state hidden input fields - if any - are not updated.

---

Summary: I think the `core.js` needs to be fixed in such way that it can ensure that `window.PrimeFaces` is a bit more persistent when `javax.faces.ViewRoot` is updated by ajax.

---

Quick work around for now for PF 11.0.0: add following script to `<h:head>` of your `500.xhtml`:

```
<script>window.PrimeFaces = PrimeFaces || { detachedWidgets: [] };</script>
```
It won't fix the failing update of `javax.faces.ViewState` but a good error page shouldn't have any JSF forms in first place.

**Comment by BalusC** [comment]
Closing off as this is not a bug in OmniFaces and a work around is available.

**Comment by melloware** [comment]
I reported it to PrimeFaces here: https://github.com/primefaces/primefaces/issues/8743

**Comment by melloware** [comment]
Problem seems to be here:

```js
 // used by @all
                else if (id === PrimeFaces.VIEW_ROOT) {

                    // backup our utils, we reset it soon
                    var ajaxUtils = PrimeFaces.ajax.Utils;

                    // reset PrimeFaces JS state because the view is completely replaced with a new one
                    window.PrimeFaces = null;

                    ajaxUtils.updateHead(content);
                    ajaxUtils.updateBody(content);
                }
```

**Comment by crick77** [comment]
Sorry for being late, I was sick and not able to work...anyway glad to see that the problem was easily reproduced. For now, BalusC's workaround works like a charm...we need to wait PF12 for a final fix.

Thank you very much!

**Comment by melloware** [comment]
thanks for reporting and yes will be fixed in 12.0.0!
